Crash in PRNGFixes.java:259

java.lang.SecurityException: Failed to read from /dev/urandom at com.microsoft.aad.adal.PRNGFixes$LinuxPRNGSecureRandom.engineNextBytes(PRNGFixes.java:259) at java.security.SecureRandom.nextBytes(SecureRandom.java:272) at com.microsoft.aad.adal.StorageHelper.encrypt(StorageHelper.java:188) at com.microsoft.aad.adal.DefaultTokenCacheStore.encrypt(DefaultTokenCacheStore.java:125) at com.microsoft.aad.adal.DefaultTokenCacheStore.setItem(DefaultTokenCacheStore.java:190) at com.microsoft.aad.adal.TokenCacheAccessor.setItemToCacheForUser(TokenCacheAccessor.java:363) at com.microsoft.aad.adal.TokenCacheAccessor.updateTokenCache(TokenCacheAccessor.java:244) at com.microsoft.aad.adal.TokenCacheAccessor.updateCachedItemWithResult(TokenCacheAccessor.java:222) at com.microsoft.aad.adal.AcquireTokenSilentHandler.acquireTokenWithCachedItem(AcquireTokenSilentHandler.java:339) at com.microsoft.aad.adal.AcquireTokenSilentHandler.tryFRT(AcquireTokenSilentHandler.java:299) at com.microsoft.aad.adal.AcquireTokenSilentHandler.tryMRRT(AcquireTokenSilentHandler.java:247) at com.microsoft.aad.adal.AcquireTokenSilentHandler.tryRT(AcquireTokenSilentHandler.java:209) at com.microsoft.aad.adal.AcquireTokenSilentHandler.getAccessToken(AcquireTokenSilentHandler.java:107) at com.microsoft.aad.adal.AcquireTokenRequest.tryAcquireTokenSilentLocally(AcquireTokenRequest.java:438) at com.microsoft.aad.adal.AcquireTokenRequest.acquireTokenSilentFlow(AcquireTokenRequest.java:401) at com.microsoft.aad.adal.AcquireTokenRequest.tryAcquireTokenSilent(AcquireTokenRequest.java:354) at com.microsoft.aad.adal.AcquireTokenRequest.performAcquireTokenRequest(AcquireTokenRequest.java:333) at com.microsoft.aad.adal.AcquireTokenRequest.access$200(AcquireTokenRequest.java:47) at com.microsoft.aad.adal.AcquireTokenRequest$1.run(AcquireTokenRequest.java:111) at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1080) at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:573) at java.lang.Thread.run(Thread.java:856)

Hi, I', also getting this exception on an API 16 device. 

```
E/AndroidRuntime: FATAL EXCEPTION: main
                  java.lang.SecurityException: Failed to read from /dev/urandom
                      at com.microsoft.aad.adal.PRNGFixes$LinuxPRNGSecureRandom.engineNextBytes(PRNGFixes.java:259)
                      at java.security.SecureRandom.nextBytes(SecureRandom.java:273)
                      at java.util.UUID.randomUUID(UUID.java:130)
                      at com.microsoft.aad.adal.AuthenticationContext.getRequestCorrelationId(AuthenticationContext.java:1062)
                      at com.microsoft.aad.adal.AuthenticationContext.acquireToken(AuthenticationContext.java:351)
                      at au.com.pacificnational.trainscanner.activities.LoginActivity$4.handleMessage(LoginActivity.java:190)
                      at android.os.Handler.dispatchMessage(Handler.java:99)
                      at android.os.Looper.loop(Looper.java:137)
                      at android.app.ActivityThread.main(ActivityThread.java:4745)
                      at java.lang.reflect.Method.invokeNative(Native Method)
                      at java.lang.reflect.Method.invoke(Method.java:511)
                      at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
                      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
                      at dalvik.system.NativeStart.main(Native Method)
                   Caused by: java.io.IOException: read failed: EBADF (Bad file number)
                      at libcore.io.IoBridge.read(IoBridge.java:442)
                      at java.io.FileInputStream.read(FileInputStream.java:179)
                      at libcore.io.Streams.readFully(Streams.java:81)
                      at java.io.DataInputStream.readFully(DataInputStream.java:120)
                      at java.io.DataInputStream.readFully(DataInputStream.java:116)
                      at com.microsoft.aad.adal.PRNGFixes$LinuxPRNGSecureRandom.engineNextBytes(PRNGFixes.java:256)
                      at java.security.SecureRandom.nextBytes(SecureRandom.java:273) 
                      at java.util.UUID.randomUUID(UUID.java:130) 
                      at com.microsoft.aad.adal.AuthenticationContext.getRequestCorrelationId(AuthenticationContext.java:1062) 
                      at com.microsoft.aad.adal.AuthenticationContext.acquireToken(AuthenticationContext.java:351) 
                      at au.com.pacificnational.trainscanner.activities.LoginActivity$4.handleMessage(LoginActivity.java:190) 
                      at android.os.Handler.dispatchMessage(Handler.java:99) 
                      at android.os.Looper.loop(Looper.java:137) 
                      at android.app.ActivityThread.main(ActivityThread.java:4745) 
                      at java.lang.reflect.Method.invokeNative(Native Method) 
                      at java.lang.reflect.Method.invoke(Method.java:511) 
                      at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786) 
                      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553) 
                      at dalvik.system.NativeStart.main(Native Method) 
                   Caused by: libcore.io.ErrnoException: read failed: EBADF (Bad file number)
                      at libcore.io.Posix.readBytes(Native Method)
                      at libcore.io.Posix.read(Posix.java:118)
                      at libcore.io.BlockGuardOs.read(BlockGuardOs.java:149)
                      at libcore.io.IoBridge.read(IoBridge.java:432)
                      at java.io.FileInputStream.read(FileInputStream.java:179) 
                      at libcore.io.Streams.readFully(Streams.java:81) 
                      at java.io.DataInputStream.readFully(DataInputStream.java:120) 
                      at java.io.DataInputStream.readFully(DataInputStream.java:116) 
                      at com.microsoft.aad.adal.PRNGFixes$LinuxPRNGSecureRandom.engineNextBytes(PRNGFixes.java:256) 
                      at java.security.SecureRandom.nextBytes(SecureRandom.java:273) 
                      at java.util.UUID.randomUUID(UUID.java:130) 
                      at com.microsoft.aad.adal.AuthenticationContext.getRequestCorrelationId(AuthenticationContext.java:1062) 
                      at com.microsoft.aad.adal.AuthenticationContext.acquireToken(AuthenticationContext.java:351) 
                      at au.com.pacificnational.trainscanner.activities.LoginActivity$4.handleMessage(LoginActivity.java:190) 
                      at android.os.Handler.dispatchMessage(Handler.java:99) 
                      at android.os.Looper.loop(Looper.java:137) 
                      at android.app.ActivityThread.main(ActivityThread.java:4745) 
                      at java.lang.reflect.Method.invokeNative(Native Method) 
                      at java.lang.reflect.Method.invoke(Method.java:511) 
                      at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786) 
                      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553) 
                      at dalvik.system.NativeStart.main(Native Method) 
```
The exception occurs after triggering ```
        mAuthContext.acquireToken(getActivity(), RESOURCE_ID, CLIENT_ID, REDIRECT_URI, PromptBehavior.Auto, getAuthInteractiveCallback());```.

Prior to that, I had called setSecretKey() via AuthenticationSettings. Can anyone assist?

--- a/adal/src/main/java/com/microsoft/aad/adal/PRNGFixes.java
+++ b/adal/src/main/java/com/microsoft/aad/adal/PRNGFixes.java
@@ -222,9 +222,11 @@ final class PRNGFixes {
 
         @Override
         protected void engineSetSeed(byte[] bytes) {
-            OutputStream out = null;
             try {
-                out = getUrandomOutputStream();
+                OutputStream out;
+                synchronized (SLOCK) {
+                    out = getUrandomOutputStream();
+                }
                 out.write(bytes);
                 out.flush();
             } catch (IOException e) {
@@ -233,13 +235,6 @@ final class PRNGFixes {
                 Logger.w(PRNGFixes.class.getSimpleName(), "Failed to mix seed into " + URANDOM_FILE);
             } finally {
                 mSeeded = true;
-                if (out != null) {
-                    try {
-                        out.close();
-                    } catch (IOException e) {
-                        Logger.v(TAG + "engineSetSeed", "Failed to close the output stream to \"/dev/urandom\" . Exception: " + e.toString());
-                    }
-                }
             }
         }
 
@@ -249,22 +244,17 @@ final class PRNGFixes {
                 // Mix in the device- and invocation-specific seed.
                 engineSetSeed(generateSeed());
             }
-            DataInputStream in = null;
+
             try {
-                in = getUrandomInputStream();
+                DataInputStream in;
                 synchronized (SLOCK) {
+                    in = getUrandomInputStream();
+                }
+                synchronized (in) {
                     in.readFully(bytes);
                 }
             } catch (IOException e) {
                 throw new SecurityException("Failed to read from " + URANDOM_FILE, e);
-            } finally {
-                if (in != null) {
-                    try {
-                        in.close();
-                    } catch (IOException e) {
-                        Logger.v(TAG + "engineNextBytes", "Failed to close the input stream to \"/dev/urandom\" . Exception: " + e.toString());
-                    }
-                }
             }
         }
 

